File: smoke_tests/ruby/deserialization.rb
# Insecure Deserialization vulnerabilities in Ruby
require 'yaml'
require 'json'
require 'marshal'

class DeserializationVulnerabilities
  def unsafe_yaml_load(user_input)
    # VULNERABLE: YAML.load can execute arbitrary code
    YAML.load(user_input)
  end

  def unsafe_marshal_load(data)
    # VULNERABLE: Marshal.load is dangerous with untrusted data
    Marshal.load(data)
  end

  def unsafe_eval(code)
    # VULNERABLE: eval with user input
    eval(code)
  end

  def unsafe_instance_eval(user_code)
    # VULNERABLE: instance_eval with user input
    instance_eval(user_code)
  end

  def unsafe_send(method_name, *args)
    # VULNERABLE: Dynamic method invocation
    send(method_name, *args)
  end

  def unsafe_constantize(class_name)
    # VULNERABLE: Arbitrary class instantiation
    class_name.constantize.new
  end

  def erb_injection(template)
    # VULNERABLE: ERB template injection
    ERB.new(template).result
  end
end

File: smoke_tests/typescript/deserialization.ts
// Insecure Deserialization vulnerabilities in TypeScript
import { serialize, deserialize } from 'some-serializer';

class DeserializationVulnerabilities {
    // VULNERABLE: Deserializing untrusted data
    deserializeUserData(data: string): any {
        return JSON.parse(data);  // Can be safe, but often misused
    }

    // VULNERABLE: eval-based deserialization
    unsafeDeserialize(serialized: string): any {
        return eval('(' + serialized + ')');
    }

    // VULNERABLE: Function constructor
    dynamicFunction(code: string): Function {
        return new Function(code);
    }

    // VULNERABLE: Using node-serialize (known vulnerable)
    nodeSerializeVuln(data: string): any {
        return deserialize(data);
    }
}
